corehq/apps/api/odata/serializers.py

```diff
@@ -71,3 +73,73 @@ def _clean_property_name(name):
                     properties.pop(property_name)
 
         return json.dumps(data, cls=DjangoJSONEncoder, sort_keys=True)
+
+
+class ODataXFormInstanceSerializer(Serializer):
+    """
+    A custom serializer that converts form data into an odata-compliant format.
+    Must be paired with ODataXFormInstanceResource
+    """
+    def to_json(self, data, options=None):
+        options = options or {}
+
+        domain = data.pop('domain', None)
+        if not domain:
+            raise Exception('API requires domain to be set! Did you add it in a custom create_response function?')
+
+        app_id = data.pop('app_id', None)
+        if not app_id:
+            raise Exception(
+                'API requires xmlns to be set! Did you add it in a custom create_response function?'
+            )
+
+        xmlns = data.pop('xmlns', None)
+        if not xmlns:
+            raise Exception(
+                'API requires xmlns to be set! Did you add it in a custom create_response function?'
+            )
+
+        api_path = data.pop('api_path', None)
+        if not api_path:
+            raise Exception(
+                'API requires api_path to be set! Did you add it in a custom create_response function?'
+            )
+
+        data = self.to_simple(data, options)
+        data['@odata.context'] = '{}#{}'.format(
+            absolute_reverse(ODataFormMetadataView.urlname, args=[domain, app_id]),
+            xmlns
+        )
+        next_url = data.pop('meta', {}).get('next')
+        if next_url:
+            data['@odata.nextLink'] = '{}{}{}'.format(get_url_base(), api_path, next_url)
+
+        # move "objects" to "value"
+        data['value'] = data.pop('objects')
+
+        form_export_schema = get_latest_form_export_schema(
+            domain, app_id, 'http://openrosa.org/formdesigner/' + xmlns
+        )
+
+        if form_export_schema:
+            export_items = [
+                item for item in form_export_schema.group_schemas[0].items
+                if isinstance(item, ExportItem)
+            ]
+
+            def _lookup(item, xform_json):
+                for path_node in item.path:
+                    try:
+                        xform_json = xform_json[path_node.name]
+                    except KeyError:
+                        return None
+                return xform_json
+
+            for i, xform_json in enumerate(data['value']):
+                data['value'][i] = {
+                    item.label.split('.')[-1].replace('#', '').replace('@', ''): _lookup(item, xform_json)
```

Comment:
this might be clearer if extracted to a function that has a useful name and tests :)